pkg/request/pool.go

```diff
@@ -67,114 +66,120 @@ type requestItem struct {
 
 // PoolOptions is the pool configuration
 type PoolOptions struct {
-	MaxSize               int
-	BatchMaxSize          int
+	MaxSize               uint64
+	BatchMaxSize          uint64
+	BatchMaxSizeBytes     uint64
+	RequestMaxBytes       uint64
 	SubmitTimeout         time.Duration
-	AutoRemoveTimeout     time.Duration
+	BatchTimeout          time.Duration
 	OnFirstStrikeTimeout  func([]byte)
 	FirstStrikeThreshold  time.Duration
+	OnSecondStrikeTimeout func()
 	SecondStrikeThreshold time.Duration
+	AutoRemoveTimeout     time.Duration
 }
 
-// NewPool constructs new requests pool
+// NewPool constructs a new requests pool
 func NewPool(log Logger, inspector RequestInspector, options PoolOptions) *Pool {
+
+	// TODO check pool options
+
 	if options.SubmitTimeout == 0 {
-		options.SubmitTimeout = defaultRequestTimeout
+		options.SubmitTimeout = defaultSubmitTimeout
+	}
+	if options.BatchTimeout == 0 {
+		options.BatchTimeout = defaultBatchTimeout
 	}
 	if options.BatchMaxSize == 0 {
 		options.BatchMaxSize = 1000
 	}
+	if options.BatchMaxSizeBytes == 0 {
+		options.BatchMaxSizeBytes = 100000
+	}
 	if options.MaxSize == 0 {
 		options.MaxSize = 10000
 	}
-
-	ps := &PendingStore{
-		Inspector:             inspector,
-		ReqIDGCInterval:       options.AutoRemoveTimeout / 4,
-		ReqIDLifetime:         options.AutoRemoveTimeout,
-		Time:                  time.NewTicker(time.Second).C,
-		StartTime:             time.Now(),
-		Logger:                log,
-		SecondStrikeThreshold: options.SecondStrikeThreshold,
-		FirstStrikeThreshold:  options.FirstStrikeThreshold,
-		Semaphore:             semaphore.NewWeighted(int64(options.MaxSize)),
-		Epoch:                 time.Second,
-		FirstStrikeCallback:   func([]byte) {},
-		SecondStrikeCallback:  func() {},
-	}
-
-	if options.OnFirstStrikeTimeout != nil {
-		ps.FirstStrikeCallback = options.OnFirstStrikeTimeout
+	if options.RequestMaxBytes == 0 {
+		options.RequestMaxBytes = defaultMaxBytes
 	}
 
-	ps.Init()
-	ps.Restart()
-
 	rp := &Pool{
-		pending:   ps,
 		logger:    log,
 		inspector: inspector,
 		semaphore: semaphore.NewWeighted(int64(options.MaxSize)),
 		options:   options,
 	}
 
-	rp.batchStore = NewBatchStore(options.MaxSize, options.BatchMaxSize, func(key string) {
-		rp.semaphore.Release(1)
-	})
+	rp.start()
 
 	return rp
 }
 
-func (rp *Pool) SetBatching(enabled bool) {
-	rp.batchStore.SetBatching(enabled)
-	if enabled {
-		atomic.StoreUint32(&rp.batchingEnabled, 1)
-	} else {
-		atomic.StoreUint32(&rp.batchingEnabled, 0)
-	}
-}
+func (rp *Pool) start() {
+	rp.lock.Lock()
```

Comment:
Not sure we have to lock here. 

We call this method from the constructor. All goroutines that can use this method see the memory writes of this function because the entire object (Pool) is never referenced by anyone before this function returns.

Reply:
Correct
Done